### PR TITLE
Add hint message when parameters is empty in clip_grad_norm_

### LINKED ISSUE #148259: Raise a warning when `torch.nn.utils.clip_grad_norm_` receives an exhausted generator
The current clip_grad_norm_ and clip_grad_value_ functions accept an Iterable[Tensor] as input. However, if that iterable is a generator that has already been consumed, then it’s effectively empty when passed in, and no gradient clipping occurs—silently. This can cause subtle bugs when the user thinks they are clipping gradients but actually aren’t.

Consider the following scenario:
```python
model = ... # some `torch.nn.Module` instance
y = model(x) # forward pass
loss = ... # calc the loss
loss.backward() # calculating gradients
params_to_clip = model.parameters() # a generator object

# e.g., iterate over the model's parameters (for some reason), unknowingly exhausting the generator
for p in params_to_clip:
    ... # do something with p

# now, the user wants to clip the gradients before doing an update step
torch.nn.utils.clip_grad_norm_(params_to_clip, ...)
# NOTE: the user unknowingly passes an empty/exhausted generator to `clip_grad_norm_` means no clipping is performed, but no warning or error is raised, leaving the user unaware.

# The user is unknowingly updating the model with unclipped gradients (since the generator was exhausted)
optimizer.step()
optimizer.zero_grad()
```

In practice, it is rare for a model to truly have zero parameters, so an empty generator is almost always a bug—often the user simply isn’t aware that model.parameters() is a generator. This can lead to unstable training if gradients aren’t actually clipped as intended.

### Alternatives

Raise a warning if clip_grad_norm_ (or clip_grad_value_) detects an exhausted generator. For instance:
```python
import types
import warnings

def clip_grad_norm_(...) -> torch.Tensor:
    if isinstance(parameters, torch.Tensor):
        parameters = [parameters]
    else:
        params_is_gen = isinstance(parameters, types.GeneratorType)
        # prevent generators from being exhausted
        parameters = list(parameters)
        if params_is_gen and len(parameters) == 0:
            warnings.warn(
                "`parameters` is an empty generator. This might be an unintended "
                "result of previous operations, meaning no gradient clipping "
                "will occur. If this is intentional, you can ignore this message."
            )
    # rest of the code remains unchanged
    ...
```

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @malfet

### PR DESCRIPTION
**Pull request by zeshengzong**
Fixes #148259

## Changes

- Add print warning message when `parameters` generator exhausted

### print warning
```python

import torch
import torch.nn as nn
import torch.optim as optim

class SimpleModel(nn.Module):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.fc = nn.Linear(10, 1)

    def forward(self, x):
        return self.fc(x)

model = SimpleModel()
criterion = nn.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

inputs = torch.randn(16, 10)
targets = torch.randn(16, 1)

outputs = model(inputs)
loss = criterion(outputs, targets)
optimizer.zero_grad()
loss.backward()

params_to_clip = model.parameters()

for p in params_to_clip:
    print(p.shape)

max_norm = 1.0
norm_type = 2.0
total_norm = nn.utils.clip_grad_norm_(params_to_clip, max_norm, norm_type)
print(f"total_norm: {total_norm}")
```

```bash
/home/zong/code/pytorch/torch/nn/utils/clip_grad.py:222: UserWarning: `parameters` is an empty generator, no gradient clipping will occur.
  warnings.warn(
total_norm: 0.0
```

### UT

```bash
pytest test/test_nn.py -k test_clip_grad_norm
```

image

### CODE REVIEW

File: torch/nn/utils/clip_grad.py [hunk 1]
```diff
@@ -213,8 +215,13 @@ def clip_grad_norm_(
     if isinstance(parameters, torch.Tensor):
         parameters = [parameters]
     else:
+        is_generator = isinstance(parameters, types.GeneratorType)
         # prevent generators from being exhausted
         parameters = list(parameters)
+        if is_generator and len(parameters) == 0:
+            warnings.warn(
+                "`parameters` is an empty generator, no gradient clipping will occur."
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply torch/nn/utils/clip_grad.py]: Level 2 will print out wrapper function, not users'

Add `stacklevel=3`, get hint below

```bash
/home/zong/code/clip.py:32: UserWarning: `parameters` is an empty generator, no gradient clipping will occur.
  total_norm = nn.utils.clip_grad_norm_(params_to_clip, max_norm, norm_type)
```

**Comment by zeshengzong** [comment]
Hello @mikaylagawarecki, please help review the change, thanks!

**Review by mikaylagawarecki — COMMENTED** [review]
Can you add a test that this warning gets raised in one of the existing clip_grad_norm tests in test/test_nn.py

**Comment by zeshengzong** [comment]
> Can you add a test that this warning gets raised in one of the existing clip_grad_norm tests in test/test_nn.py

Test added, thanks! @mikaylagawarecki

**Review by jbschlosser — APPROVED** [review]
